File: Lowest Common Ancestor of a Binary Search Tree.cpp
#include <iostream>
using namespace std; 

struct TreeNode {
int val;
 TreeNode *left;
 TreeNode *right;
 TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  };

int max(TreeNode* p,TreeNode* q){
	if(p->val>q->val)
		return p->val;
	return q->val;
}

int min(TreeNode* p,TreeNode* q){
	if(p->val<q->val)
		return p->val;
	return q->val;
}

TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
	if(root==NULL || p==NULL || q==NULL)
		retur NULL;
	else if(root->val < max(p,q) && root->val > min(p,q))
		return root;
	else if(root->val > max(p,q))
		return lowestCommonAncestor(root->left,p,q);
	else if(root->val < min(p,q))
		return lowestCommonAncestor(root->right,p,q);
	else if(root->val == p->val)
		return p;
	else if(root ->val == q->val)
		return q;
}

File: Lowest Common Ancestor of a Binary Search Tree.py
class TreeNode(object):
	def __init__(self, x):
		self.val = x
        self.left = None
        self.right = None

def maxNode(p,q):
	if p.val>q.val:
		return p.val
	return q.val


def minNode(p,q):
	if p.val<q.val:
		return p.val
	return q.val


class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
    	if root==None or p==None or q==None:
    		return None
    	elif root.val<maxNode(p,q) and root.val>minNode(p,q):
    		return root
    	elif root.val>maxNode(p,q):
    		return self.lowestCommonAncestor(root.left,p,q)
    	elif root.val<minNode(p,q):
    		return self.lowestCommonAncestor(root.right,p,q)
    	elif root.val==p.val:
    		return p
    	elif root.val==q.val:
    		return q